src/super_context.ml

```diff
@@ -865,53 +865,120 @@ module PP = struct
+             [ Build.run ~context:sctx.context (Ok ppx_exe) args
+             ; Build.create_file digest_path
+             ])
+      )
+
   (* Generate rules to build the .pp files and return a new module map where all filenames
      point to the .pp files *)
-  let pped_modules sctx ~dir ~dep_kind ~modules ~preprocess ~preprocessor_deps ~lib_name
-        ~scope =
+  let pped_linted_modules sctx ~dir ~dep_kind ~modules ~lint ~preprocess
+        ~preprocessor_deps ~lib_name ~scope =
     let preprocessor_deps =
       Build.memoize "preprocessor deps"
         (Deps.interpret sctx ~scope ~dir preprocessor_deps)
     in
+    let lint_module = lint_module sctx ~dir ~dep_kind ~lint ~lib_name ~scope in
     String_map.map modules ~f:(fun (m : Module.t) ->
-      match Preprocess_map.find m.name preprocess with
-      | No_preprocessing -> setup_reason_rules sctx ~dir m
-      | Action action ->
-        pped_module m ~dir ~f:(fun _kind src dst ->
-          add_rule sctx
-            (preprocessor_deps
-             >>>
-             Build.path src
-             >>^ (fun _ -> [src])
-             >>>
-             Action.run sctx
-               (Redirect
-                  (Stdout,
-                   target_var,
-                   Chdir (root_var,
-                          action)))
-               ~dir
-               ~dep_kind
-               ~targets:(Static [dst])
-               ~scope))
-        |> setup_reason_rules sctx ~dir
-      | Pps { pps; flags } ->
-        let ppx_exe = get_ppx_driver sctx pps ~dir ~dep_kind in
-        let m = setup_reason_rules sctx ~dir m in
-        pped_module m ~dir ~f:(fun kind src dst ->
-          add_rule sctx
-            (preprocessor_deps
-             >>>
-             Build.run ~context:sctx.context
-               (Ok ppx_exe)
-               [ As flags
-               ; A "--dump-ast"
-               ; As (cookie_library_name lib_name)
-               ; A "-o"; Target dst
-               ; Ml_kind.ppx_driver_flag kind; Dep src
-               ])
-        )
+        match Preprocess_map.find m.name preprocess with
+        | No_preprocessing ->
+          let ast = setup_reason_rules sctx ~dir m in
+          lint_module ~ast:(lazy ast) ~source:m;
+          ast
+        | Action action ->
+          let ast =
+            pped_module m ~dir ~f:(fun _kind src dst ->
+              add_rule sctx
+                (preprocessor_deps
+                >>>
+                Build.path src
+                >>^ (fun _ -> [src])
+                >>>
+                Action.run sctx
+                  (Redirect
+                      (Stdout,
+                      target_var,
+                      Chdir (root_var,
+                              action)))
+                  ~dir
+                  ~dep_kind
+                  ~targets:(Static [dst])
+                  ~scope))
+            |> setup_reason_rules sctx ~dir in
+          lint_module ~ast:(lazy (setup_reason_rules sctx ~dir m)) ~source:m;
```

Comment:
Shouldn't we pass just `ast` instead of `(lazy (...))`? It's possible that the file won't parse with `refmt` if we need a preprocessor

Reply:
It will be weird because now we are going to lint the post processed code only with ppx linters. While action linters will work on the original source.

This case is for the situation where we have a preprocessor with action but the linter is ppx. I'm not really sure what's desirable here...

Reply:
> It will be weird because now we are going to lint the post processed code only with ppx linters. While action linters will work on the original source.

I agree this is surprising, but if we want to make it more homogeneous, we should apply action linters after action preprocessors.

Reply:
I think that will defeat the purpose of some linters. For example, I'm thinking of something like ocp-indent as a linter. That should clearly run on the unprocessed code. Why do you think that what we have here isn't consistent? Sure it doesn't work for all cases (e.g. you're preprocessing reason and linting it with ppx) but it is consistent in the sense that it only works on the original source.

Reply:
What would really be nice here is if you could specify explicitly when the linter would run (before/after) preprocessing.

Reply:
But ocp-indent doesn't support cpp, cppo, etc... so running it before the pre-processor is not very useful either.

In the end, most OCaml code now doesn't use a preprocessor or uses a ppx rewriter. The last use of preprocessors is probably conditional compilation and we have a pure ppx replacement at Jane Street. For the cases where the lint field doesn't work, one can always use an explicit `(alias ...)` stanza.

Reply:
I think making ocp-indent work with cppo isn't very hard. But OK, I pushed a commit to just pass `ast` now.